### PR TITLE
Serverextension

### PR DESCRIPTION
**Pull request by ian-r-rose**
There is more to do, but this is getting big enough as it is, so I'd prefer to do it in follow-up PRs. This creates a server extension for managing dask clusters. It allows you to list, start and stop them via a REST api. A UI for this is implemented in the side panel. You can also inject code to connect to any of the clusters via right-clicking on their list item

image

TODOS.
- [x] Add scaling functionality to the REST API.
- [x] Figure out tornado IOLoop isues (see discussion in #30).
- [ ] Add drag/drop for client connection code snippets. This will require some upstream JupyterLab enhancements.
- [x] Add a proxy to the server extension to get around CORS issues.
- [x] Figure out a reasonable config API on the server side for configuring what kind of cluster to start.

### CODE REVIEW

**Comment by mrocklin** [comment]
I've had some difficulty in setting this up.  Help would be appreciated.

<details>

```
mrocklin@carbon:~/workspace/dask-labextension$ conda create -n dev jupyterlab nodejs dask bokeh ipython
Solving environment: done
	
## Package Plan ##

  environment location: /home/mrocklin/Software/anaconda/envs/dev

  added / updated specs: 
    - bokeh
    - dask
    - ipython
    - jupyterlab
    - nodejs


The following NEW packages will be INSTALLED:

    backcall:          0.1.0-py37_0           
    blas:              1.0-mkl                
    bleach:            3.0.2-py37_0           
    bokeh:             1.0.1-py37_0           
    ca-certificates:   2018.03.07-0           
    certifi:           2018.10.15-py37_0      
    click:             7.0-py37_0             
    cloudpickle:       0.6.1-py37_0           
    cytoolz:           0.9.0.1-py37h14c3975_1 
    dask:              0.20.0-py37_0          
    dask-core:         0.20.0-py37_0          
    decorator:         4.3.0-py37_0           
    distributed:       1.24.0-py37_0          
    entrypoints:       0.2.3-py37_2           
    freetype:          2.9.1-h8a8886c_1       
    gmp:               6.1.2-h6c8ec71_1       
    heapdict:          1.0.0-py37_2           
    intel-openmp:      2019.0-118             
    ipykernel:         5.1.0-py37h39e3cac_0   
    ipython:           7.1.1-py37h39e3cac_0   
    ipython_genutils:  0.2.0-py37_0           
    jedi:              0.13.1-py37_0          
    jinja2:            2.10-py37_0            
    jpeg:              9b-h024ee3a_2          
    jsonschema:        2.6.0-py37_0           
    jupyter_client:    5.2.3-py37_0           
    jupyter_core:      4.4.0-py37_0           
    jupyterlab:        0.35.3-py37_0          
    jupyterlab_server: 0.2.0-py37_0           
    libedit:           3.1.20170329-h6b74fdf_2
    libffi:            3.2.1-hd88cf55_4       
    libgcc-ng:         8.2.0-hdf63c60_1       
    libgfortran-ng:    7.3.0-hdf63c60_0       
    libpng:            1.6.35-hbc83047_0      
    libsodium:         1.0.16-h1bed415_0      
    libstdcxx-ng:      8.2.0-hdf63c60_1       
    libtiff:           4.0.9-he85c1e1_2       
    locket:            0.2.0-py37_1           
    markupsafe:        1.0-py37h14c3975_1     
    mistune:           0.8.4-py37h7b6447c_0   
    mkl:               2019.0-118             
    mkl_fft:           1.0.6-py37h7dd41cf_0   
    mkl_random:        1.0.1-py37h4414c95_1   
    msgpack-python:    0.5.6-py37h6bb024c_1   
    nbconvert:         5.3.1-py37_0           
    nbformat:          4.4.0-py37_0           
    ncurses:           6.1-hf484d3e_0         
    nodejs:            8.12.0-he6710b0_0      
    notebook:          5.7.0-py37_0           
    numpy:             1.15.3-py37h1d66e8a_0  
    numpy-base:        1.15.3-py37h81de0dd_0  
    olefile:           0.46-py37_0            
    openssl:           1.1.1-h7b6447c_0       
    packaging:         18.0-py37_0            
    pandas:            0.23.4-py37h04863e7_0  
    pandoc:            2.2.3.2-0              
    pandocfilters:     1.4.2-py37_1           
    parso:             0.3.1-py37_0           
    partd:             0.3.9-py37_0           
    pexpect:           4.6.0-py37_0           
    pickleshare:       0.7.5-py37_0           
    pillow:            5.3.0-py37h34e0f95_0   
    pip:               10.0.1-py37_0          
    prometheus_client: 0.4.2-py37_0           
    prompt_toolkit:    2.0.7-py37_0           
    psutil:            5.4.8-py37h7b6447c_0   
    ptyprocess:        0.6.0-py37_0           
    pygments:          2.2.0-py37_0           
    pyparsing:         2.2.2-py37_0           
    python:            3.7.1-h0371630_3       
    python-dateutil:   2.7.5-py37_0           
    pytz:              2018.7-py37_0          
    pyyaml:            3.13-py37h14c3975_0    
    pyzmq:             17.1.2-py37h14c3975_0  
    readline:          7.0-h7b6447c_5         
    send2trash:        1.5.0-py37_0           
    setuptools:        40.5.0-py37_0          
    six:               1.11.0-py37_1          
    sortedcontainers:  2.0.5-py37_0           
    sqlite:            3.25.2-h7b6447c_0      
    tblib:             1.3.2-py37_0           
    terminado:         0.8.1-py37_1           
    testpath:          0.4.2-py37_0           
    tk:                8.6.8-hbc83047_0       
    toolz:             0.9.0-py37_0           
    tornado:           5.1.1-py37h7b6447c_0   
    traitlets:         4.3.2-py37_0           
    wcwidth:           0.1.7-py37_0           
    webencodings:      0.5.1-py37_1           
    wheel:             0.32.2-py37_0          
    xz:                5.2.4-h14c3975_4       
    yaml:              0.1.7-had09818_2       
    zeromq:            4.2.5-hf484d3e_1       
    zict:              0.1.3-py37_0           
    zlib:              1.2.11-ha838bed_2      

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate dev
#
# To deactivate an active environment, use
#
#     $ conda deactivate

mrocklin@carbon:~/workspace/dask-labextension$ conda activate dev
mrocklin@carbon:~/workspace/dask-labextension$ git branch
  canavandl/individual_routes
  canavandl/jupyterlab_0.7.0_compat
  license
  master
  mega-update
  release-0.1.0
* serverextension
mrocklin@carbon:~/workspace/dask-labextension$ git pull ian-r-rose serverextension 
From github.com:ian-r-rose/dask-labextension
 * branch            serverextension -> FETCH_HEAD
Already up to date.
mrocklin@carbon:~/workspace/dask-labextension$ pip install -e -q . 
-q should either be a path to a local project or a VCS url beginning with svn+, git+, hg+, or bzr+
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
mrocklin@carbon:~/workspace/dask-labextension$ pip install -e . 
Obtaining file:///home/mrocklin/workspace/dask-labextension
Requirement already satisfied: notebook in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from dask-labextension==0.1.0) (5.7.0)
Requirement already satisfied: traitlets>=4.2.1 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (4.3.2)
Requirement already satisfied: ipython-genutils in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (0.2.0)
Requirement already satisfied: jupyter-core>=4.4.0 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (4.4.0)
Requirement already satisfied: tornado>=4 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (5.1.1)
Requirement already satisfied: pyzmq>=17 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (17.1.2)
Requirement already satisfied: prometheus-client in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (0.4.2)
Requirement already satisfied: Send2Trash in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (1.5.0)
Requirement already satisfied: jinja2 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (2.10)
Requirement already satisfied: terminado>=0.8.1 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (0.8.1)
Requirement already satisfied: ipykernel in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (5.1.0)
Requirement already satisfied: nbformat in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (4.4.0)
Requirement already satisfied: nbconvert in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (5.3.1)
Requirement already satisfied: jupyter-client>=5.2.0 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from notebook->dask-labextension==0.1.0) (5.2.3)
Requirement already satisfied: decorator in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from traitlets>=4.2.1->notebook->dask-labextension==0.1.0) (4.3.0)
Requirement already satisfied: six in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from traitlets>=4.2.1->notebook->dask-labextension==0.1.0) (1.11.0)
Requirement already satisfied: MarkupSafe>=0.23 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from jinja2->notebook->dask-labextension==0.1.0) (1.0)
Requirement already satisfied: ipython>=5.0.0 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipykernel->notebook->dask-labextension==0.1.0) (7.1.1)
Requirement already satisfied: jsonschema!=2.5.0,>=2.4 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbformat->notebook->dask-labextension==0.1.0) (2.6.0)
Requirement already satisfied: bleach in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (3.0.2)
Requirement already satisfied: testpath in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (0.4.2)
Requirement already satisfied: pandocfilters>=1.4.1 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (1.4.2)
Requirement already satisfied: pygments in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (2.2.0)
Requirement already satisfied: entrypoints>=0.2.2 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (0.2.3)
Requirement already satisfied: mistune>=0.7.4 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from nbconvert->notebook->dask-labextension==0.1.0) (0.8.4)
Requirement already satisfied: python-dateutil>=2.1 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from jupyter-client>=5.2.0->notebook->dask-labextension==0.1.0) (2.7.5)
Requirement already satisfied: pickleshare in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.7.5)
Requirement already satisfied: backcall in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.1.0)
Requirement already satisfied: jedi>=0.10 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.13.1)
Requirement already satisfied: pexpect; sys_platform != "win32" in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (4.6.0)
Requirement already satisfied: setuptools>=18.5 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (40.5.0)
Requirement already satisfied: prompt-toolkit<2.1.0,>=2.0.0 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (2.0.7)
Requirement already satisfied: webencodings in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from bleach->nbconvert->notebook->dask-labextension==0.1.0) (0.5.1)
Requirement already satisfied: parso>=0.3.0 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from jedi>=0.10->ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.3.1)
Requirement already satisfied: ptyprocess>=0.5 in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from pexpect; sys_platform != "win32"->ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.6.0)
Requirement already satisfied: wcwidth in /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages (from prompt-toolkit<2.1.0,>=2.0.0->ipython>=5.0.0->ipykernel->notebook->dask-labextension==0.1.0) (0.1.7)
mkl-random 1.0.1 requires cython, which is not installed.
Installing collected packages: dask-labextension
  Running setup.py develop for dask-labextension
Successfully installed dask-labextension
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
mrocklin@carbon:~/workspace/dask-labextension$ jupyter serverextension enable --sys-prefix dask_labextension
Enabling: dask_labextension
- Writing config: /home/mrocklin/Software/anaconda/envs/dev/etc/jupyter
    - Validating...
      dask_labextension 0.1.0 OK
mrocklin@carbon:~/workspace/dask-labextension$ jupyter lab
[I 06:16:31.377 LabApp] JupyterLab extension loaded from /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages/jupyterlab
[I 06:16:31.377 LabApp] JupyterLab application directory is /home/mrocklin/Software/anaconda/envs/dev/share/jupyter/lab
[W 06:16:31.378 LabApp] JupyterLab server extension not enabled, manually loading...
[I 06:16:31.380 LabApp] JupyterLab extension loaded from /home/mrocklin/Software/anaconda/envs/dev/lib/python3.7/site-packages/jupyterlab
[I 06:16:31.380 LabApp] JupyterLab application directory is /home/mrocklin/Software/anaconda/envs/dev/share/jupyter/lab
[I 06:16:31.381 LabApp] Serving notebooks from local directory: /home/mrocklin/workspace/dask-labextension
[I 06:16:31.381 LabApp] The Jupyter Notebook is running at:
[I 06:16:31.381 LabApp] http://localhost:8888/
[I 06:16:31.381 LabApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[I 06:16:33.527 LabApp] Node v8.12.0
    
[I 06:16:33.530 LabApp] Build is up to date

```

</details>

image

**Comment by ian-r-rose** [comment]
As far as I can tell, you have not installed the client-side part of the extension. From the root of the repository
```bash
jlpm install
jlpm build
jupyter labextension install .
```
With a server extension the install story for users is going to be a bit more complicated (though we can talk through some steps to mitigate that), so we had better make it worth while :)

**Comment by mrocklin** [comment]
OK, I run the following:

```
jlpm install
jlpm build
jupyter labextension install .
pip install -e .
jupyter serverextension enable --sys-prefix dask_labextension
```

Things show up nicely in JLab, but pressing the "new cluster" button results in a 404

```
[W 13:30:52.231 LabApp] 404 PUT /dask?1541266252211 (127.0.0.1) 7.58ms referer=http://localhost:8888/lab
[W 13:30:52.379 LabApp] 404 PUT /dask?1541266252359 (127.0.0.1) 9.07ms referer=http://localhost:8888/lab
```

**Comment by ian-r-rose** [comment]
Hmm. Does `jupyter serverextension list` indicate that the server side component is installed and importable?

**Comment by ian-r-rose** [comment]
When I enter `jupyter serverextension list`:
```
config dir: /home/ian/anaconda3/envs/jupyter/etc/jupyter
    nbresuse  enabled 
    - Validating...
      nbresuse  OK
    jupyterlab  enabled 
    - Validating...
      jupyterlab 1.0.0a0 OK
    jupyterlab_latex  enabled 
    - Validating...
      jupyterlab_latex 0.4.1 OK
    jupyterlab_github  enabled 
    - Validating...
      jupyterlab_github 0.7.0 OK
    dask_labextension  enabled 
    - Validating...
      dask_labextension 0.1.0 OK
```

**Comment by mrocklin** [comment]
```
mrocklin@carbon:~/workspace/dask-labextension$ jupyter serverextension enable --sys-prefix dask_labextension
Enabling: dask_labextension
- Writing config: /home/mrocklin/Software/anaconda/etc/jupyter
    - Validating...
      dask_labextension  OK
mrocklin@carbon:~/workspace/dask-labextension$ jupyter serverextension list
config dir: /home/mrocklin/.jupyter
    dask_labextension disabled
    - Validating...
      dask_labextension  OK
config dir: /home/mrocklin/Software/anaconda/etc/jupyter
    jupyterlab  enabled 
    - Validating...
      jupyterlab  OK
    nbdime  enabled 
    - Validating...
      nbdime  OK
    nbserverproxy  enabled 
    - Validating...
      nbserverproxy  OK
    dask_labextension  enabled 
    - Validating...
      dask_labextension  OK
```

**Comment by ian-r-rose** [comment]
Hm, I see `dask_labextension` twice in that listing. Once enabled, and once disabled. Maybe go into the `/home/mrocklin/.jupyter/jupyter_notebook_config.json` file and manually remove that one?

**Comment by mrocklin** [comment]
That works.  Everything seems pretty slick to me so far.  I'll take a look at the web server side code next.

**Comment by mrocklin** [comment]
I suspect that you have plenty of plans here @ian-r-rose .  The two that strike me most immediately are the following:

1.  Improving the ease with which users can create a client that points to a particular cluster
2.  Improving the ability of users to scale their clusters up and down 

cc @rabernat and @jacobtomlinson , who I think may also have thoughts here and may also want to experiment a bit

**Comment by ian-r-rose** [comment]
Some quick responses:
1. Did you play with the client-code-injection functionality? I also have some plans for drag/drop of client code that points to a cluster (the interaction being something like "drag the cluster list item to a notebook"). That, however, is pending on some upstream work in JupyterLab: jupyterlab/jupyterlab#5571.
2. I think this could be done with the REST API via a PATCH operation. However, I found it a bit clunky on the server side to introspect a cluster to find its current scale, and whether it is adaptive. Am I missing something there?

**Comment by mrocklin** [comment]
> Did you play with the client-code-injection functionality (in the right-click menu)? 

Ah, I didn't even realize that this existed.  Pretty slick.  

> I think this could be done with the REST API via a PATCH operation. However, I found it a bit clunky on the server side to introspect a cluster to find its current scale, and whether it is adaptive. Am I missing something there?

You shouldn't have to introspect any state here.  You just have to call `cluster.scale(n)`.  The cluster will look at its current state and take the appropriate measures.  Same with `cluster.adapt(minimum=..., maximum=...)`

**Comment by ian-r-rose** [comment]
But I think I do want at least *some* introspection to show to the user on the frontend. That is, wouldn't the user want to know the current scaling (or adaptive settings) when deciding whether/how to scale up or down?

**Comment by mrocklin** [comment]
Yeah, fair enough.  Short term you might want the following:

https://github.com/dask/distributed/blob/ab2835271fff6182605deb3f443c25f425b72699/distributed/deploy/cluster.py#L109-L136

**Comment by ian-r-rose** [comment]
Okay, now this has a first pass at the ability to rescale a cluster, both adaptively and statically:
image

**Comment by mrocklin** [comment]
Sorry for the delay.  Was travelling yesterday.  In principle this seems great.  I'm going to recommend a couple of style changes if they're easy to do.

-  We include the total cores and memory in the text defining each cluster (presumably this requires periodically checking in with the server side?)
-  Clicking on a cluster makes it take over the iframe'd plots.  We should then be able to remove the link button
-  Is it possible to inline the functionality in the dialog directly within the sidebar?  For example maybe clicking on a cluster entry also makes that entry in the clusters list slightly taller, and includes a text field to scale the number of target workers directly.  Perhaps that also has an `adapt` button that switches the entry to min/max

    ```
    Cluster 1: 20 cores 100GB   <close button>
      cores [ 20 ] + / -        <adapt button>
    ```

I'm just saying what I think might be nice though.  I don't have a good cost model for how expensive this is to actually do.

Also cc'ing @jcrist on this PR.  He was just asking about something like this for dask-yarn on Amazon's Elastic MapReduce

**Comment by ian-r-rose** [comment]
On Tue, Nov 6, 2018 at 11:35 AM Matthew Rocklin <notifications@github.com>
wrote:

> Sorry for the delay. Was travelling yesterday. In principle this seems
> great. I'm going to recommend a couple of style changes if they're easy to
> do.
>
No worries!

>
>    -
>
>    We include the total cores and memory in the text defining each
>    cluster (presumably this requires periodically checking in with the server
>    side?)
>
> Yes, this information can be included with little effort. I am currently
refreshing the listing every 5 seconds (we can adjust that number as
necessary).

>
>    -
>
>    Clicking on a cluster makes it take over the iframe'd plots. We should
>    then be able to remove the link button
>
> Yep, this is a nice idea.

>
>    -
>
>    Is it possible to inline the functionality in the dialog directly
>    within the sidebar? For example maybe clicking on a cluster entry also
>    makes that entry in the clusters list slightly taller, and includes a text
>    field to scale the number of target workers directly. Perhaps that also has
>    an adapt button that switches the entry to min/max
>
>    Cluster 1: 20 cores 100GB   <close button>
>      cores [ 20 ] + / -        <adapt button>
>
>
> This is doable, but with a bit more work. Perhaps we can move the dialog
content into the sidebar in a follow-up? Regarding adapt: I am currently
caching the `Adaptive` object returned by the `cluster.adapt()` function in
order to check the minimum and maximum number of workers when reporting to
the frontend. I just realized, however, that this would miss out-of-band
adjustments to the scaling. That is, if the user scales a cluster in the
notebook, the side panel wouldn't see that. What would you think about
exposing the current `adaptive` as a public property? Or we could check for
the private `_adaptive` while begging forgiveness from the public API
deities.

**Comment by mrocklin** [comment]
> I just realized, however, that this would miss out-of-band
adjustments to the scaling. That is, if the user scales a cluster in the
notebook, the side panel wouldn't see that. 

I think that as we're doing things now the user doesn't have direct access to the cluster object from the notebook.  They would have to get decently clever to change things.  

> What would you think about
exposing the current `adaptive` as a public property? Or we could check for
the private `_adaptive` while begging forgiveness from the public API
deities.

I think that you should go ahead and be as dangerous as you like here.  I suspect that this whole part of Dask is likely to undergo fairly heavy churn in the next few months.  A lot of people are currently pushing on parts of this space.  I think that for the near-term future we should develop on cutting edge, and we'll clean things up and stabilize after things settle down.  

Having features like what is in this PR really helps to inform other development.

**Comment by mrocklin** [comment]
> This is doable, but with a bit more work. Perhaps we can move the dialog
content into the sidebar in a follow-up?

Sounds good to me

**Comment by ian-r-rose** [comment]
I've just added a proxy for the locally hosted cluster dashboards. This should go a long way towards removing CORS errors when using this extension (cf #23). It is based off of a modified version of `nbserverproxy`. This also avoids the need to check for the static image (at least when the clusters are managed by the extension).

**Comment by ian-r-rose** [comment]
At this point I'm fairly satisfied with where this stands. The biggest blocker I see is resolving the tornado event loop issues.

**Comment by mrocklin** [comment]
@minrk do you know why adding the following line in a server extension would cause the Jupyter server to hang?

    asyncio.set_event_loop_policy(tornado.platform.asyncio.AnyThreadEventLoopPolicy())

**Comment by mrocklin** [comment]
I'll also look into how we might extend this in configuration to support, the KubeCluster, SGECluster, and so on.

**Comment by mrocklin** [comment]
I'll try to make some time for this today.

**Comment by mrocklin** [comment]
OK, I've added enough config handling to make this work well enough with dask-kubernetes, dask-jobqueue, dask-yarn and such in 393e75a8ef21ef128f102411296a1bcf43f8734b

I also did some small style change with mypy (I'm new to this) and flake8.  Feel free to disregard if preferred.

**Comment by mrocklin** [comment]
I've raised the event loop issue upstream at https://github.com/jupyter/jupyter/issues/378 .  I'll continue investigating locally.

**Comment by ian-r-rose** [comment]
> OK, I've added enough config handling to make this work well enough with dask-kubernetes, dask-jobqueue, dask-yarn and such in 393e75a8ef21ef128f102411296a1bcf43f8734b

Awesome, that is exactly what I was hoping you would do with the cluster factory.

> I also did some small style change with mypy (I'm new to this) and flake8.  Feel free to disregard if preferred.

I am also new to mypy, but I really like the type hinting in TypeScript, so thought I'd try it out here. It doesn't feel quite as well thought out, but maybe I am using it non-optimally.

I had been formatting with black, but don't really have strong opinions on that front. flake8 is fine.

**Comment by mrocklin** [comment]
Also @minrk and I are discussing the event loop stuff in the Pangeo gitter chat starting here: https://gitter.im/pangeo-data/Lobby?at=5be455864e7ca1452073dbae

**Comment by mrocklin** [comment]
Also, another UI suggestion that arose while I was using this a bit more.  When injecting the code into the notebook it might be nicer to inject a new cell just above the active cell rather than insert it into the current cell.  Injecting into the current one can fail if that cell is markdown, or if the user is in an inconvenient location.  I think that injecting just above will almost always be welcome.

**Comment by ian-r-rose** [comment]
Sure, this is a nice idea.

**Comment by mrocklin** [comment]
OK, so my plan here is as follows:

1.  Release the `dask/distributed` library with the necessary fix (should hopefully happen later today)
2.  Merge this
3.  Try using this in the dask-examples binder, or possibly one of the Pangeo deployments and learn a bit
4.  Release this

For the binder deployments I suspect that we'll want a way to launch a cluster or two automatically at startup, probably through configuration.  I think that that should be pretty easy to do though, and should probably be separate from this PR regardless.

**Comment by ian-r-rose** [comment]
Sounds good. When you release the new `distributed` I can specify that in the `setup.py` here. I also added a stub to the README about configuring the cluster factory that I was hoping you could expand on a bit.

We can certainly launch an initial cluster via configuration. It should also be possible to do it via the REST API (either in a startup script or in the client-side code at JupyterLab startup).

File: dask_labextension/manager.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            An optional string id for the cluster. If not given, a random id
+            will be chosen.
+
+        Returns
+        cluster_model : a dask cluster model.
+        """
+        if not cluster_id:
+            cluster_id = str(uuid4())
+        cluster = self._cluster_factory()
+        self._n_clusters = self._n_clusters + 1
```
Comment (dhirschfeld) [inline dask_labextension/manager.py]:
```suggestion
        self._n_clusters += 1
```

Comment (ian-r-rose) [reply dask_labextension/manager.py]: Thanks @dhirschfeld! Too many languages rattling around my head.

Comment (dhirschfeld) [reply dask_labextension/manager.py]: I think the new suggested changes are fantastic, so I figured I'd use one there. 

It's not much, but I had to do a double take there to make sure it was _just_ adding 1. I thought maybe there's a reason it's not using the short-form - i.e. it was a similarly spelt but different variable on the rhs. Using the augmented assignment operators avoids this ambiguity.

**Comment by dhirschfeld** [comment]
I was just perusing to see if I could make use of this new functionality.

I'm not sure I can since I've got a custom dask cluster on Windows HPC. I'm guessing I'd need to inherit from [`Cluster`](https://github.com/dask/distributed/blob/cb10d6b10ad7800648b44ab3b194220d1e6fc5f8/distributed/deploy/cluster.py#L14) for this to work? Is that all?

**Comment by ian-r-rose** [comment]
Our intention is to make this work for a bunch of different kinds of Dask clusters, though we are still working through the details of what that means. So if you are interested in trying to get this working in your Windows HPC environment, your experience would be quite valuable to the extension development.

You are right that your own custom cluster should implement the Cluster interface (whether or not it explicitly inherits from it). You would then try to specify a cluster factory function using the Dask config system, which would allow the cluster manager to start new ones on command. There may be some changes we have to make here for that to be a smooth process, so do let us know if you decide to try it!

**Comment by mrocklin** [comment]
@ian-r-rose I'd like to try this with dask-kubernetes on a pangeo binder deployment.  Is there an easy way to install a Jupyter labextension from a github repository?  I'm looking for a command to place into my postBuild script.

I took a look at https://jupyterlab.readthedocs.io/en/stable/user/extensions.html but it only talks about installing from npm or local.  Presumably I could clone this repository and build locally?

**Comment by ian-r-rose** [comment]
Briefly, as I'm on the road: the best option should be to clone it and then
do a local install, as you suggest.

**Comment by jhamman** [comment]
I'm currently testing this in https://github.com/pangeo-data/pangeo-example-notebooks/pull/28

**Comment by jhamman** [comment]
@ian-r-rose - any chance we can get this wrapped up and into a release this week? I'd love to use the updated extension in a series of demos/tutorials next week.

**Comment by mrocklin** [comment]
I suspect that @ian-r-rose is waiting for us to try things out and verify
that it functions correctly.  If that's done, then yes, we should probably
merge and release.

**Comment by ian-r-rose** [comment]
I had to set this aside for a few days due to the JupyterLab developer meeting. I think we can get this merged and released this week!

I want to fix some of the rough edges you found @jhamman, and maybe move some of the information in the hover text for clusters directly into the listing UI. Once that is set, I think we should be good to go. Any more feedback from testing in the Pangeo environment would be greatly appreciated, of course.

**Comment by jhamman** [comment]
Thanks @ian-r-rose - The binder repo liked to above is still working with this extension and I'll be happy to run through things again once you've taken another pass.

**Comment by ian-r-rose** [comment]
@jhamman, this is close-to-ready for another round of testing. I know you have a bunch of free time at AGU!

**Comment by ian-r-rose** [comment]
@jhamman, good point. I've added another button to the cluster listing for injecting the client code. I've also fleshed out the context menu a bit, so the user can manipulate clusters using either avenue.

**Comment by consideRatio** [comment]
Exciting work!! 👍

**Comment by jhamman** [comment]
@ian-r-rose - I'm happy here for now. Is there more you'd like to sneak in here?

**Comment by ian-r-rose** [comment]
I think this is ready from my perspective. Thanks for putting it through the paces @jhamman!

**Comment by ian-r-rose** [comment]
Okay @mrocklin, I think this is just about ready. I'm sure stuff is busy right now, but if you get the chance to take this for a spin...

**Comment by mrocklin** [comment]
Spin taken.  In general I think it looks great.  The only-speedbump I ran into was not knowing immediately how to connect a notebook to it.  The `< >` button ended up being correct, but I wonder if we might replace this text with something like "Connect".

I'd be more than happy to merge this in now though if you'd like to move on.  I think that users will be able to figure that out.

**Comment by ian-r-rose** [comment]
We could certainly change it to `CONNECT`. I was feeling that the button panel was getting kind of cramped, and hoped that having a descriptive hover text would be sufficient (I was also considering restoring the up/down arrows instead of `SCALE`). But overall it's not a huge deal.

This PR is getting pretty big as it is, so I think having more refinements in follow-ups is a good idea. If you are happy, I'd say merge away!

**Comment by mrocklin** [comment]
I'll give @jhamman an hour to look things over, and then plan to merge afterwards.

**Comment by mrocklin** [comment]
This is in.  Thanks @ian-r-rose . 

Should we release?

**Comment by ian-r-rose** [comment]
Can we merge without squashing? This is a big PR and represents a lot of work, and most of the commits have semantic meaning.

**Comment by mrocklin** [comment]
Ah, my apologies.  I just squash-and-release by default these days.  It's been soon enough that rewriting history seems reasonable to me.  Care to just force push to master?

**Comment by ian-r-rose** [comment]
Sure, thanks!

**Comment by ian-r-rose** [comment]
@mrocklin can you add me as a maintainer of the `dask-labextension` package on PyPI?

**Comment by mrocklin** [comment]
Sure, what is your PyPI username?

**Comment by ian-r-rose** [comment]
Username `irose`

**Comment by mrocklin** [comment]
Done

**Comment by jhamman** [comment]
Thanks @ian-r-rose!

**Comment by ian-r-rose** [comment]
Okay, published with server extension `dask_labextension` 3.1, and lab extension `dask-labextension@0.3.0`. The python package needed a quick patch release to correctly include the config yaml.

**Pull request merged**